src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -1842,9 +1843,79 @@ public override BoundNode VisitConditionalAccess(BoundConditionalAccess node)
+                    !conversion.IsUserDefined)
+                {
+                    // where T : U
+                    // T ?? U or U ?? T
+
+                    // T ?? U
+                    // If T is nullable, U is also nullable - result is nullable
+                    // If T is not nullable - result is nullable if U is nullable
+                    // If U is nullable - result is nullable
+                    // If U is not nullable, T is also not nullable - result is not nullable
+                    // So, nullability of the result can be inferred from nullability of U, and the type of the result is U.   
+
+                    // U ?? T
+                    // If U is nullable - result is nullable
+                    // If U is not nullable, T is also not nullable - result is not nullable
+                    // If T is nullable, U is also nullable - result is nullable
+                    // If T is not nullable - result is nullable if U is nullable
+                    // So, nullability of the result can be inferred from nullability of U, but the type of the result is T.   
+                    // At the moment we don't have a way to represent this correlation, result type is one type parameter, but
+                    // nullability comes from another. 
+                    // Ideally, we would want to have the following behavior:
+                    //     U x = U ?? T - no warning
+                    //     T y = U ?? T - a warning
+                    // But we can track the state only in the way when either both produce a warning, or none.
+                    // It feels like it is reasonable to prefer the latter approach, i.e. produce no warnings 
+                    // for both scenarios - no false diagnostics.
+                    resultAnnotation = NullableAnnotation.NotNullable; // Inherit nullability of U
+                }
+                else
+                {
+                    resultAnnotation = NullableAnnotation.NullableBasedOnAnalysis;
+                }
+            }
+            else
+            {
+                NullableAnnotation receiverAnnotation = receiverType.GetValueNullableAnnotation();
+                NullableAnnotation accessAnnotation = _resultType.GetValueNullableAnnotation();
+                if (receiverAnnotation.IsAnyNullable() || accessAnnotation.IsAnyNullable())
+                {
+                    resultAnnotation = NullableAnnotation.NullableBasedOnAnalysis;
+                }
+                else if (receiverAnnotation == NullableAnnotation.Unknown || accessAnnotation == NullableAnnotation.Unknown)
+                {
+                    resultAnnotation = NullableAnnotation.Unknown;
+                }
+                else
+                {
+                    resultAnnotation = NullableAnnotation.NotNullableBasedOnAnalysis;
+                }
+            }
```

Comment:
>} [](start = 12, length = 1)

Could you update the speclet to document the handling of null-coalescing-operator?

I'm still reading through this block, but it makes me uneasy. I wonder whether it could be expressed in terms of some existing or re-usable abstractions.

Reply:
You are correct in that the speclet does not cover every single operator. That's because some operators have (or had) straightforward implementations.
But that is no longer the case here, which is why I'm asking that we spec this.

---
In reply to: [236460582](https://github.com/dotnet/roslyn/pull/31264#discussion_r236460582) [](ancestors = 236460582,236459591)

Reply:
> But that is no longer the case here, which is why I'm asking that we spec this.

I see no reason to spec this because there is nothing special about this operator. Nullability is inferred based on its semanics, I see no good reason to duplicate all this conditional logic in the speclet, this is just an implementation.

---
In reply to: [236466167](https://github.com/dotnet/roslyn/pull/31264#discussion_r236466167) [](ancestors = 236466167,236460582,236459591)

Reply:
> there is nothing special about this operator

The previous code was `receiverType.IsNullable | _resultType.IsNullable`. Can we agree that it changed significantly and that it is no longer straightforward?

>  I see no good reason to duplicate all this conditional logic in the speclet

I'm not asking to duplicate this logic (I agree that would not help). 
I'm asking to describe the behavior in some abstraction if possible. If there is no abstraction beyond the detailed logic in the code, then we should discuss with Mads since he will have to specify the behavior for some or all operators at some point.

---
In reply to: [236466863](https://github.com/dotnet/roslyn/pull/31264#discussion_r236466863) [](ancestors = 236466863,236466167,236460582,236459591)

Reply:
> I'm asking to describe the behavior in some abstraction if possible. 

The behavior is fully defined by semantics of conditional access, this is the same as for any other similar operator. If you feel like this semantics should be re-described (language spec does it in all the details), I will add this as a comment preceding the code. I see no good reason for this information to be in the speclet.

---
In reply to: [236781403](https://github.com/dotnet/roslyn/pull/31264#discussion_r236781403) [](ancestors = 236781403,236466863,236466167,236460582,236459591)